cmd/cluster-node-tuning-operator/main.go

```diff
@@ -1,64 +1,149 @@
 	"github.com/openshift/cluster-node-tuning-operator/version"
 )
 
+const (
+	operandFilename  = "openshift-tuned"
+	operatorFilename = "cluster-node-tuning-operator"
+	metricsHost      = "0.0.0.0"
+	leaderElectionID = "node-tuning-operator-lock"
+)
+
 var (
-	boolVersion = flag.Bool("version", false, "show program version and exit")
-	boolLocal   = flag.Bool("local", false, "local run outside a pod")
+	scheme = apiruntime.NewScheme()
 )
 
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(tunedv1.AddToScheme(scheme))
+	utilruntime.Must(mcov1.AddToScheme(scheme))
+	utilruntime.Must(apiconfigv1.Install(scheme))
+}
+
 func printVersion() {
 	klog.Infof("Go Version: %s", runtime.Version())
 	klog.Infof("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH)
 	klog.Infof("%s Version: %s", tunedv1.TunedClusterOperatorResourceName, version.Version)
 }
 
 func main() {
-	const (
-		operandFilename  string = "openshift-tuned"
-		operatorFilename string = "cluster-node-tuning-operator"
-	)
+	var boolVersion bool
+	var enableLeaderElection bool
+	flag.BoolVar(&boolVersion, "version", false, "show program version and exit")
+	flag.BoolVar(&enableLeaderElection, "enable-leader-election", true,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
 
 	runAs := filepath.Base(os.Args[0])
 
-	stopCh := signals.SetupSignalHandler()
-
 	switch runAs {
 	case operatorFilename:
+		var metricsAddr string
+		flag.StringVar(&metricsAddr, "metrics-addr", fmt.Sprintf("%s:%d", metricsHost, metrics.MetricsPort),
+			"The address the metric endpoint binds to.")
+
 		klog.InitFlags(nil)
 		flag.Parse()
 
 		printVersion()
 
-		if *boolVersion {
+		if boolVersion {
 			os.Exit(0)
 		}
 
-		go metrics.RunServer(metrics.MetricsPort, stopCh)
-		metrics.RegisterVersion(version.Version)
-		controller, err := operator.NewController()
+		// We have two namespaces that we need to watch:
+		// 1. NTO namespace - for NTO resources
+		// 2. None namespace - for cluster wide resources
+		ntoNamespace := config.OperatorNamespace()
+		namespaces := []string{
+			ntoNamespace,
+			metav1.NamespaceNone,
+		}
+
+		restConfig := ctrl.GetConfigOrDie()
+		le := util.GetLeaderElectionConfig(restConfig, enableLeaderElection)
+		mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+			NewCache:                cache.MultiNamespacedCacheBuilder(namespaces),
+			Scheme:                  scheme,
+			LeaderElection:          true,
+			LeaderElectionID:        leaderElectionID,
+			LeaderElectionNamespace: ntoNamespace,
+			LeaseDuration:           &le.LeaseDuration.Duration,
+			RetryPeriod:             &le.RetryPeriod.Duration,
+			RenewDeadline:           &le.RenewDeadline.Duration,
+			Namespace:               ntoNamespace,
+		})
+
 		if err != nil {
-			klog.Fatal(err)
+			klog.Exit(err)
 		}
 
-		controller.BecomeLeader(stopCh)
+		mgr.Add(metrics.Server{})
```

Comment:
The class names where kept to have some sort of a diff (at least the controller)
But now I intend to change the names to:
reconciler.go -> cvo_controller.go
controller.go -> tuned_controller.go
The struct names will be changed accordingly to be more clear.

More detailed comments will be added in main to explain that, but in general
cvo_controller - reconciles requests for CVO objects. It owns (default) tuned and deamonset it creates.
It also watches for all other tuned objects and profiles in order to update the status. 

tuned_controller - reconciles requests for tuned objects and performs the logic for syncing profiles/nodes/rendered tuned in correlation with machine configs. 

I agree that there would be some more cleanups to make the code more clear and move/remove some code sections from one controller or the other.
I do think that having 2 reconciles / controllers should be the form of implementation here but I'm open for changes
(cc @cynepco3hahue)

Reply:
> The class names were kept to have some sort of a diff (at least the controller) but now I intend to change the names to: reconciler.go -> cvo_controller.go controller.go -> tuned_controller.go The struct names will be changed accordingly to be more clear.

It is still not clear to me why we need two controllers, but I'm not against it as it might aid modularity and a clear separation of tasks.  But the "tasks" need to be clear, agreed and make sense.  If we really do need two (or even more controllers), then perhaps having a `controller` prefix to separate them would be better.  E.g. `controller_whatever.go`.  But `controller_cvo.go` still seems like a strange name and misleading.  We need to come up with a better name if we do split it.  I do not have any better suggestions at this point.

Reply:
In general, the assumption of the controller is that the controller should work only on one type of resource. Like you said it is better from a modularity and readability perspective. So the idea is to have:
1. operator controller to deploy the default set of resources(default tuned and default daemon set)
2. controller to handle the behavior around the tuned resources.